### PR TITLE
[SPARK-15591] [WEBUI] Paginate Stage Table in Stages tab

### PR DESCRIPTION
**Pull request by nblintao**
## What changes were proposed in this pull request?

This patch adds pagination support for the Stage Tables in the Stage tab. Pagination is provided for all of the four Job Tables (active, pending, completed, and failed). Besides, the paged stage tables are also used in JobPage (the detail page for one job) and PoolPage.

Interactions (jumping, sorting, and setting page size) for paged tables are also included.
## How was this patch tested?

Tested manually by using checking the Web UI after completing and failing hundreds of jobs.  Same as the testings for [Paginate Job Table in Jobs tab](https://github.com/apache/spark/pull/13620).

This shows the pagination for completed stages:
paged stage table

### CODE REVIEW

**Comment by ajbozarth** [comment]
I haven't been able to look at the code yet but I checked it out and ran it and found a bug. When you change one table (completed or failed) it resets the other table. I'll also check if this happens on your jobs page pr as well

**Comment by nblintao** [comment]
@ajbozarth Thanks for pointing it out. It is true for both the stages page and the job page. It doesn't keep track of the status of other tables when one table is changed.

I will try to fix it recently.

**Comment by nblintao** [comment]
@ajbozarth Could you please try again to see whether or not have my new commit fixed the problem you mentioned above?
Also, regarding the bug you mentioned in #13620, are the _stage_ tables working properly in the history servers?
Thanks!

**Comment by ajbozarth** [comment]
Thanks, I'll check it out when I have a chance today. And I'll double check the history server, but I'm pretty sure it was both.

**Comment by nblintao** [comment]
I have fixed some linking bugs in stage tables (including the history server problem reported by @ajbozarth ). I have also prevented displaying empty tables in pool pages.
I have tested the links for the AllStagesPage and the JobPage in both the normal UI and the history UI, and tested the PoolPage in the normal UI.
Please have a review of my code and check if it works properly. I will fix the corresponding bugs in #13620 after this PR is merged.

**Comment by ajbozarth** [comment]
I finally had a chance to checkout and test your updates and it looks great on both Web UI and History server (I did't check Pool Page because I've never used it). One small thing though, how are the >> and << arrows supposed to work? Because my assumption would be that they go to the first or last page overall, but you have them implemented to skip 10 pages forward or backward. Just wondering if this was your intent and if so why?

**Comment by nblintao** [comment]
@ajbozarth Thanks. That was implemented in the pageNavigation of the PagedTable by @zsxwing from PR #7399. I didn't change it, and it seems useful to me. It works like a pagination for the page navigator itself, since the navigator displays 10 links at once.

**Comment by nblintao** [comment]
Jenkins, retest this please

**Comment by ajbozarth** [comment]
@nblintao thanks for that PR link, that had a great explanation for why he decided it should work that way.

I haven't looked at the code yet, but from a UI perspective this is great.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala [hunk 1]
```diff
@@ -38,22 +38,20 @@ private[ui] class AllStagesPage(parent: StagesTab) extends WebUIPage("") {
...
-          parent.basePath, parent.progressListener, isFairScheduler = parent.isFairScheduler,
-          killEnabled = parent.killEnabled)
+        new StageTableBase(request, activeStages, "activeStage", parent.basePath, subPath,
+          parent.progressListener, parent.isFairScheduler, parent.killEnabled, false)
       val pendingStagesTable =
-        new StageTableBase(pendingStages.sortBy(_.submissionTime).reverse,
-          parent.basePath, parent.progressListener, isFairScheduler = parent.isFairScheduler,
-          killEnabled = false)
+        new StageTableBase(request, pendingStages, "pendingStage", parent.basePath, subPath,
+          parent.progressListener, parent.isFairScheduler, false, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala]: nit: could you add the parameter names back to improve the readability?

File: core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala [hunk 2]
```diff
@@ -38,22 +38,20 @@ private[ui] class AllStagesPage(parent: StagesTab) extends WebUIPage("") {
...
       val completedStagesTable =
-        new StageTableBase(completedStages.sortBy(_.submissionTime).reverse, parent.basePath,
-          parent.progressListener, isFairScheduler = parent.isFairScheduler, killEnabled = false)
+        new StageTableBase(request, completedStages, "completedStage", parent.basePath, subPath,
+          parent.progressListener, parent.isFairScheduler, false, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala]: nit: add the parameter names

File: core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala [hunk 3]
```diff
@@ -38,22 +38,20 @@ private[ui] class AllStagesPage(parent: StagesTab) extends WebUIPage("") {
...
       val failedStagesTable =
-        new FailedStageTable(failedStages.sortBy(_.submissionTime).reverse, parent.basePath,
-          parent.progressListener, isFairScheduler = parent.isFairScheduler)
+        new StageTableBase(request, failedStages, "failedStage", parent.basePath, subPath,
+          parent.progressListener, parent.isFairScheduler, false, true)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllStagesPage.scala]: nit: add the parameter names

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 4]
```diff
@@ -229,20 +229,20 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
...
-          parent.basePath, parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler,
-          killEnabled = parent.killEnabled)
+        new StageTableBase(request, activeStages, "activeStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, parent.killEnabled, false)
       val pendingOrSkippedStagesTable =
-        new StageTableBase(pendingOrSkippedStages.sortBy(_.stageId).reverse,
-          parent.basePath, parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler,
-          killEnabled = false)
+        new StageTableBase(request, pendingOrSkippedStages, "pendingStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, false, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: nit: add the parameter names for `false`s.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 5]
```diff
@@ -229,20 +229,20 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
...
       val completedStagesTable =
-        new StageTableBase(completedStages.sortBy(_.submissionTime).reverse, parent.basePath,
-          parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler, killEnabled = false)
+        new StageTableBase(request, completedStages, "completedStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, false, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: nit: add the parameter names for `false`s.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 6]
```diff
@@ -229,20 +229,20 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
...
       }
 
+      val basePath = "jobs/job"
+
       val activeStagesTable =
-        new StageTableBase(activeStages.sortBy(_.submissionTime).reverse,
-          parent.basePath, parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler,
-          killEnabled = parent.killEnabled)
+        new StageTableBase(request, activeStages, "activeStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, parent.killEnabled, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: nit: add the parameter name for `false`.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 7]
```diff
@@ -229,20 +229,20 @@ private[ui] class JobPage(parent: JobsTab) extends WebUIPage("job") {
...
       val completedStagesTable =
-        new StageTableBase(completedStages.sortBy(_.submissionTime).reverse, parent.basePath,
-          parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler, killEnabled = false)
+        new StageTableBase(request, completedStages, "completedStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, false, false)
       val failedStagesTable =
-        new FailedStageTable(failedStages.sortBy(_.submissionTime).reverse, parent.basePath,
-          parent.jobProgresslistener, isFairScheduler = parent.isFairScheduler)
+        new StageTableBase(request, failedStages, "failedStage", parent.basePath,
+          basePath, parent.jobProgresslistener, parent.isFairScheduler, false, true)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: nit: add the parameter names for `false` and `true`.

File: core/src/main/scala/org/apache/spark/ui/jobs/PoolPage.scala [hunk 8]
```diff
@@ -42,19 +42,21 @@ private[ui] class PoolPage(parent: StagesTab) extends WebUIPage("pool") {
         case Some(s) => s.values.toSeq
         case None => Seq[StageInfo]()
       }
-      val activeStagesTable = new StageTableBase(activeStages.sortBy(_.submissionTime).reverse,
-        parent.basePath, parent.progressListener, isFairScheduler = parent.isFairScheduler,
-        killEnabled = parent.killEnabled)
+      val shouldShowActiveStages = activeStages.nonEmpty
+      val activeStagesTable =
+        new StageTableBase(request, activeStages, "activeStage", parent.basePath, "stages/pool",
+          parent.progressListener, parent.isFairScheduler, parent.killEnabled, false)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/PoolPage.scala]: nit: add the parameter name for `false`

File: core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala [hunk 9]
```diff
@@ -17,61 +17,323 @@
...
+  override def goButtonFormPath: String = {
+    val encodedSortColumn = URLEncoder.encode(sortColumn, "UTF-8")
+    s"$parameterPath&$stageTag.sort=$encodedSortColumn&$stageTag.desc=$desc"
   }
 
-  def toNodeSeq: Seq[Node] = {
-    listener.synchronized {
-      stageTable(renderStageRow, stages)
+  override def headers: Seq[Node] = {
+    val stageHeadersAndCssClasses: Seq[(String, Option[(String, Boolean)], Boolean)] =
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/jobs/StageTable.scala]: nit: could you add a comment to explain the meaning of `(String, Option[(String, Boolean)], Boolean)`?

**Comment by zsxwing** [comment]
@nblintao thanks for your contribution. Looks pretty good except some nits. Could you fix them, please?

**Comment by nblintao** [comment]
Thanks, @zsxwing . I have fixed them in the commit above.

**Comment by zsxwing** [comment]
LGTM. @nblintao Thanks. Merging to master.

**Pull request closed without merging**